### DISCUSSION TITLE
Accessing DRED Data for TDR debugging

### DISCUSSION THREAD

**Discussion by coingod** [amerkoleci/Vortice.Windows#358] [Q&A] [answered]
Hello, I'm trying to implement DRED Support in my application to debug TDRs, as described in:
https://microsoft.github.io/DirectX-Specs/d3d/DeviceRemovedExtendedData.html

I'm able to enable the feature and query DRED once a Device Removal occurs.

```
// Turn on AutoBreadcrumbs and Page Fault reporting (DRED 1.1)
var dredSettings = D3D12.D3D12GetDebugInterface<ID3D12DeviceRemovedExtendedDataSettings>();
dredSettings.SetAutoBreadcrumbsEnablement(DredEnablement.ForcedOn);
dredSettings.SetPageFaultEnablement(DredEnablement.ForcedOn);

//...

// Device removal has been detected...
var dred = D3DDevice.QueryInterface<ID3D12DeviceRemovedExtendedData1>();
var dredAutoBreadcrumbsOutput = dred.AutoBreadcrumbsOutput;
var dredPageFaultOutput = dred.PageFaultAllocationOutput;

IntPtr nodePtr = dredAutoBreadcrumbsOutput.HeadAutoBreadcrumbNode;
```

What I'm not sure is how can I access the `AutoBreadcrumbNode` struct from the pointer returned by `dredAutoBreadcrumbsOutput.HeadAutoBreadcrumbNode`.
I would like to transverse this linked list of nodes, but I need to get the data from that pointer.

I tried doing the following:
```AutoBreadcrumbNode node = Marshal.PtrToStructure<AutoBreadcrumbNode>(nodePtr);```

But it fails in runtime with the following error:
`System.InvalidCastException: 'Unable to cast object of type 'System.__ComObject' to type 'Vortice.Direct3D12.ID3D12GraphicsCommandList'.'`

What would be the correct way of reading the AutoBreadcrumbNodes? Is it supported?

**Comment by amerkoleci** [comment, marked as answer]
Hi,
Can you share some piece of C++ code on how you would do that?

**Reply by coingod** [reply]
Based on the documentation linked above, I was thinking something like this:
```
        ComPtr<ID3D12DeviceRemovedExtendedData1> pDred;
        m_device->QueryInterface(IID_PPV_ARGS(&pDred));

        D3D12_DRED_AUTO_BREADCRUMBS_OUTPUT1 DredAutoBreadcrumbsOutput;
        D3D12_DRED_PAGE_FAULT_OUTPUT DredPageFaultOutput;
        pDred->GetAutoBreadcrumbsOutput1(&DredAutoBreadcrumbsOutput);
        pDred->GetPageFaultAllocationOutput(&DredPageFaultOutput);

        // Pointer to the next node in the list or nullptr if this is the last node
        const D3D12_AUTO_BREADCRUMB_NODE1* node = DredAutoBreadcrumbsOutput.pHeadAutoBreadcrumbNode;
        while (node != nullptr)
        {
            // Log node information to console...
            DEBUG_LOG(node->pCommandListDebugNameA);
            DEBUG_LOG(node->pCommandQueueDebugNameA);

            int opCount = node->BreadcrumbCount;
            for (size_t i = 0; i < opCount; i++)
            {
                DEBUG_LOG(node->pCommandHistory + sizeof(D3D12_AUTO_BREADCRUMB_OP) * i);
            }

            // Move to next node
            node = node->pNext;
        }
```

**Reply by amerkoleci** [reply]
There are marshal issues on SharpGen side, will fix it ASAP and let you know once done, thanks for bringing this issue!

**Reply by amerkoleci** [reply]
Here we go, you can see an example here:

https://github.com/amerkoleci/Vortice.Windows/blob/main/src/samples/HelloDirect3D12/D3D12GraphicsDevice.cs#L444
